### PR TITLE
[eks/cluster] Update to use AWS Auth API

### PR DESCRIPTION
**Pull request by Nuru**
## Breaking Changes

>[!WARNING]
> This release contains breaking changes to the `eks/cluster` component.
Read the migration guidance in the [CHANGELOG](./modules/eks/cluster/CHANGELOG.md)

## what

- Upgrade the `eks/cluster` component to use the AWS API for access control

## why

- The old mechanism, using a ConfigMap, was unreliable

## references

- `terraform-aws-eks-cluster` [v4.0.0 Release Notes](https://github.com/cloudposse/terraform-aws-eks-cluster/releases/tag/4.0.0)
- `terraform-aws-eks-cluster` [v4 Migration Guide](https://github.com/cloudposse/terraform-aws-eks-cluster/blob/main/docs/migration-v3-v4.md)
- [New API for EKS access control](https://github.com/aws/containers-roadmap/issues/185#issuecomment-1863025784)

### CODE REVIEW

File: modules/eks/cluster/CHANGELOG.md [hunk 1]
```diff
@@ -20,13 +220,14 @@ single node, removing the high availability that comes from having a node per Av
...
-our ServiceControlPolicy `DenyEC2NonNitroInstances`, `DenyNonNitroInstances` (obsolete, replaced by
-`DenyEC2NonNitroInstances`), and/or `DenyEC2InstancesWithoutEncryptionInTransit`, you will want to update them to
-v0.12.0 or choose a difference instance type.
+`c7a.medium`) in each of 3 Availability Zones. This will provide the compute power needed to initialize add-ons, and
+will provide high availability for the cluster. As a bonus, it will also remove the need to deploy Karpenter to Fargate.
+
+**NOTE about instance type**: The `c7a.medium` instance type is relatively new. If you have deployed an old version of
+our [ServiceControlPolicy](https://github.com/cloudposse/terraform-aws-service-control-policies)
+`DenyEC2NonNitroInstances`, `DenyNonNitroInstances` (obsolete, replaced by `DenyEC2NonNitroInstances`), and/or
+`DenyEC2InstancesWithoutEncryptionInTransit`, you will want to update them to v0.14.1 or choose a difference instance
```
Comment (dudymas) [inline modules/eks/cluster/CHANGELOG.md:229]:
```suggestion
`DenyEC2InstancesWithoutEncryptionInTransit`, you will want to update them to v0.14.1 or choose a different instance
```

File: modules/eks/cluster/main.tf [hunk 2]
```diff
@@ -10,48 +10,75 @@ locals {
...
   karpenter_role_arn = one(aws_iam_role.karpenter[*].arn)
 
-  worker_role_arns = compact(concat(
+  linux_worker_role_arns = compact(concat(
     var.map_additional_worker_roles,
-    local.managed_worker_role_arns,
-    [local.karpenter_role_arn]
+    # As of Karpenter v0.35.0, there is no entry in the official Karpenter documentation
+    # stating how to configure Karpenter node roles via EKS Access Entries.
+    # However, it is launching unmanaged worker nodes, so it makes sense they they
```
Comment (dudymas) [inline modules/eks/cluster/main.tf:66]:
```suggestion
    # However, it is launching unmanaged worker nodes, so it makes sense that they
```

File: modules/eks/cluster/main.tf [hunk 3]
```diff
@@ -10,48 +10,75 @@ locals {
...
+    # be configured as EC2_LINUX unmanaged worker nodes. Of course, this does probably
```
Comment (dudymas) [inline modules/eks/cluster/main.tf:67]:
```suggestion
    # be configured as EC2_LINUX unmanaged worker nodes. Of course, this probably
```

File: modules/eks/cluster/variables.tf [hunk 4]
```diff
@@ -93,27 +93,11 @@ variable "cluster_log_retention_period" {
...
-}
-
-variable "map_additional_worker_roles" {
-  type        = list(string)
-  description = "AWS IAM Role ARNs of worker nodes to add to `aws-auth` ConfigMap"
-  default     = []
-  nullable    = false
-}
-
+# TODO:
```
Comment (dudymas) [inline modules/eks/cluster/variables.tf:96]: have we made tickets for these?

File: modules/eks/cluster/main.tf [hunk 5]
```diff
@@ -10,48 +10,75 @@ locals {
...
+    }
+  }
 
-  # Existing managed worker role ARNs
-  managed_worker_role_arns = coalesce(local.eks_outputs.eks_managed_node_workers_role_arns, [])
+  access_entry_map = merge(local.aws_team_roles_access_entry_map, local.aws_sso_access_entry_map, local.iam_roles_access_entry_map, local.iam_users_access_entry_map)
 
-  # If Karpenter IAM role is enabled, add it to the `aws-auth` ConfigMap to allow the nodes launched by Karpenter to join the EKS cluster
+  # Existing Fargate Profile role ARNs
+  fargate_profile_role_arns = local.eks_outputs.fargate_profile_role_arns
```
Comment (dudymas) [inline modules/eks/cluster/main.tf:57]: per tflint, this is no longer in use
```suggestion
```

File: modules/eks/cluster/CHANGELOG.md [hunk 6]
```diff
@@ -1,11 +1,211 @@
...
+
+With the fixes included and AWS Terraform Provider v5.43.0 and Karpenter v0.33.0, the
+`legacy_do_not_create_karpenter_instance_profile` is now obsolete. After upgrading both this component and the
+`eks/karpenter` component, if you had it in your configuration, you can remove it. If you had previously set it to
+`false`, removing it may cause an error when you apply the changes. If you see an error about the
+`aws_iam_instance_profile` resource being destroyed (cannot be destroyed because it is in use, has dependencies, and/or
+has role attached), you can simply remove the resource from the Terraform state with `[atmos] terraform state rm`,
+because it will be managed by the Karpenter controller instead of Terraform.
+
+### Migration Procedure
```
Comment (dudymas) [inline modules/eks/cluster/CHANGELOG.md:44]: Assuming this file grows arbitrarily in the future, assume there could be a collision with this title:
```suggestion
### Access Control API Migration Procedure
```

File: modules/eks/cluster/CHANGELOG.md [hunk 7]
```diff
@@ -1,11 +1,211 @@
...
+
+  and run `atmos terraform apply` again. This will cause EKS to ignore the `aws-auth` ConfigMap, but will not remove it.
+  Again, this will cause a lot of IAM changes due to the potential for the EKS OIDC thumbprint to change, but this is
+  not a problem.
+
+- Manually remove the `aws-auth` ConfigMap. You can do this with
+  `kubectl delete configmap aws-auth --namespace kube-system`. This will not affect the cluster, because it is now being
+  managed by the new access control API, but it will reduce the possibility of confusion in the future.
+
+### End of Migration
```
Comment (dudymas) [inline modules/eks/cluster/CHANGELOG.md:195]: Assuming this file grows arbitrarily in the future, assume there could be a collision with this title:
```suggestion
### End of Access Control API Migration
```

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -44,9 +44,11 @@ repos:
         name: prettier
         entry: prettier --write --prose-wrap always --print-width 120
         types: ["markdown"]
+        # Excluding modules/eks/cluster/README.md because it chokes on the output
+        # of the `terraform-docs` command for that file.
         exclude: |
           (?x)^(
-              README.md |
+              README.md | modules/eks/cluster/README.md |
```
Comment (milldr) [inline .pre-commit-config.yaml:51]: We don't want to ignore any specific file here. Instead, wrap the section you want ignored in the README directly.

See Slack thread respond and this ARB
https://github.com/orgs/cloudposse-corp/discussions/233

**Review by milldr — DISMISSED** [review]
please see comment about prettier

File: modules/eks/cluster/README.md [hunk 9]
```diff
@@ -73,40 +153,86 @@ components:
...
+            # We recommend no autoscaling for the main node group, so it will
+            # stay at the specified desired group size, with additional
+            # capacity provided by Karpenter. Nevertheless, we recommend
+            # deploying enough capacity in the node group to handle your
+            # baseload requirements, and in production, we recommend you
+            # have a large enough node group to handle 3/2 (1.5) times your
+            # baseload requirements, to handle the loss of a single AZ.
+            desired_group_size: 3 # number of instances to start with, should be >= number of AZs
+            min_group_size: 3 # must be  >= number of AZs
+            max_group_size: 3
```
Comment (dudymas) [inline modules/eks/cluster/README.md:175]: This have been a point confusion for multiple end users. If you have 3 AZ's, setting this to 5 or to 3 is effectively rendered the same and has no effect. This is due to [this line in the node group module](https://github.com/cloudposse/terraform-aws-components/blob/main/modules/eks/cluster/modules/node_group_by_region/main.tf#L13-L15)

Really, we should change this variable to be more clear, like `desired_per_az` and `min_per_az`. 

```suggestion
            desired_group_size: 3 # number of instances to start with, should be a multiple of the # of AZs you have
            min_group_size: 3 # must be a multiple of the # of AZs you have
            max_group_size: 3 # must be a multiple of the # of AZs you have
```

File: modules/eks/cluster/variables.tf [hunk 10]
```diff
@@ -93,27 +93,11 @@ variable "cluster_log_retention_period" {
...
-  nullable    = false
-}
-
-variable "map_additional_worker_roles" {
-  type        = list(string)
-  description = "AWS IAM Role ARNs of worker nodes to add to `aws-auth` ConfigMap"
-  default     = []
-  nullable    = false
-}
-
```
Comment (dudymas) [inline modules/eks/cluster/variables.tf:116]: The variable `map_additional_worker_roles` is still used in your main.tf, so this needs to be restored

File: .pre-commit-config.yaml (hunk 8 above)
Comment (dudymas) [reply .pre-commit-config.yaml]: I've pushed a fix for this now

**Review by dudymas — CHANGES_REQUESTED** [review]
Please look at the issue with `map_additional_worker_roles`. It was removed from `variables.tf` but is still in use in `main.tf`. I believe @Benbentwo last touched that variable, so he might have more context around its usage/intent.

File: modules/eks/cluster/variables.tf (hunk 4 above)
Comment (Nuru) [reply modules/eks/cluster/variables.tf:96]: No. We will do it when we need it. So far, none of our customers need it.

File: modules/eks/cluster/README.md (hunk 9 above)
Comment (Nuru) [reply modules/eks/cluster/README.md:175]: @dudymas You are mistaken. 

When you use 3 AZs:
- If you set `max_group_size` to 3, you get 3 node groups/ASGs (one per AZ), each with a `max_size` of 1
- If you set `max_group_size` to 5, you get 3 node groups/ASGs (one per AZ), 1 of which has a `max_size` of 1, while the other 2 have `max_size` of 2

**Comment by Nuru** [comment]
> Please look at the issue with `map_additional_worker_roles`. It was removed from `variables.tf` but is still in use in `main.tf`. I believe @Benbentwo last touched that variable, so he might have more context around its usage/intent.

I had neglected to add the new `variables-deprecated.tf` file, which is where I had moved that variable to. Fixed now.

**Review by milldr — APPROVED** [review]

File: modules/eks/cluster/README.md (hunk 9 above)
Comment (dudymas) [reply modules/eks/cluster/README.md:175]: Desk-checked this again. You're right, and I missed the way index was getting added. As for the behavior in the past, it must have been a different module I'm getting this confused with.

**Review by dudymas — APPROVED** [review]
LGTM

**Pull request merged**